// @tsparticles/interaction-external-bounce/cjs/Bouncer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bouncer = void 0;
const engine_1 = require("@tsparticles/engine");
const Bounce_js_1 = require("./Options/Classes/Bounce.js");
const bounceMode = "bounce", squareExp = 2, double = 2, half = 0.5, halfPI = Math.PI * half, toleranceFactor = 10, minRadius = 0;
class Bouncer extends engine_1.ExternalInteractorBase {
    constructor(container) {
        super(container);
        this._processBounce = (position, radius, area) => {
            const query = this.container.particles.quadTree.query(area, (p) => this.isEnabled(p));
            for (const particle of query) {
                if (area instanceof engine_1.Circle) {
                    (0, engine_1.circleBounce)((0, engine_1.circleBounceDataFromParticle)(particle), {
                        position,
                        radius,
                        mass: radius ** squareExp * halfPI,
                        velocity: engine_1.Vector.origin,
                        factor: engine_1.Vector.origin,
                    });
                }
                else if (area instanceof engine_1.Rectangle) {
                    (0, engine_1.rectBounce)(particle, (0, engine_1.calculateBounds)(position, radius));
                }
            }
        };
        this._processMouseBounce = () => {
            const container = this.container, pxRatio = container.retina.pixelRatio, tolerance = toleranceFactor * pxRatio, mousePos = container.interactivity.mouse.position, radius = container.retina.bounceModeDistance;
            if (!radius || radius < minRadius || !mousePos) {
                return;
            }
            this._processBounce(mousePos, radius, new engine_1.Circle(mousePos.x, mousePos.y, radius + tolerance));
        };
        this._singleSelectorBounce = (selector, div) => {
            const container = this.container, query = document.querySelectorAll(selector);
            if (!query.length) {
                return;
            }
            query.forEach((item) => {
                const elem = item, pxRatio = container.retina.pixelRatio, pos = {
                    x: (elem.offsetLeft + elem.offsetWidth * half) * pxRatio,
                    y: (elem.offsetTop + elem.offsetHeight * half) * pxRatio,
                }, radius = elem.offsetWidth * half * pxRatio, tolerance = toleranceFactor * pxRatio, area = div.type === "circle"
                    ? new engine_1.Circle(pos.x, pos.y, radius + tolerance)
                    : new engine_1.Rectangle(elem.offsetLeft * pxRatio - tolerance, elem.offsetTop * pxRatio - tolerance, elem.offsetWidth * pxRatio + tolerance * double, elem.offsetHeight * pxRatio + tolerance * double);
                this._processBounce(pos, radius, area);
            });
        };
    }
    clear() {
    }
    init() {
        const container = this.container, bounce = container.actualOptions.interactivity.modes.bounce;
        if (!bounce) {
            return;
        }
        container.retina.bounceModeDistance = bounce.distance * container.retina.pixelRatio;
    }
    async interact() {
        const container = this.container, options = container.actualOptions, events = options.interactivity.events, mouseMoveStatus = container.interactivity.status === engine_1.mouseMoveEvent, hoverEnabled = events.onHover.enable, hoverMode = events.onHover.mode, divs = events.onDiv;
        if (mouseMoveStatus && hoverEnabled && (0, engine_1.isInArray)(bounceMode, hoverMode)) {
            this._processMouseBounce();
        }
        else {
            (0, engine_1.divModeExecute)(bounceMode, divs, (selector, div) => this._singleSelectorBounce(selector, div));
        }
        await Promise.resolve();
    }
    isEnabled(particle) {
        const container = this.container, options = container.actualOptions, mouse = container.interactivity.mouse, events = (particle?.interactivity ?? options.interactivity).events, divs = events.onDiv;
        return ((!!mouse.position && events.onHover.enable && (0, engine_1.isInArray)(bounceMode, events.onHover.mode)) ||
            (0, engine_1.isDivModeEnabled)(bounceMode, divs));
    }
    loadModeOptions(options, ...sources) {
        if (!options.bounce) {
            options.bounce = new Bounce_js_1.Bounce();
        }
        for (const source of sources) {
            options.bounce.load(source?.bounce);
        }
    }
    reset() {
    }
}
exports.Bouncer = Bouncer;
