// @tsparticles/shape-circle/cjs/CircleDrawer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CircleDrawer = void 0;
const engine_1 = require("@tsparticles/engine");
const double = 2, doublePI = Math.PI * double, sides = 12, maxAngle = 360, minAngle = 0, origin = { x: 0, y: 0 };
class CircleDrawer {
    draw(data) {
        const { context, particle, radius } = data;
        if (!particle.circleRange) {
            particle.circleRange = { min: minAngle, max: doublePI };
        }
        const circleRange = particle.circleRange;
        context.arc(origin.x, origin.y, radius, circleRange.min, circleRange.max, false);
    }
    getSidesCount() {
        return sides;
    }
    particleInit(container, particle) {
        const shapeData = particle.shapeData, angle = shapeData?.angle ?? {
            max: maxAngle,
            min: minAngle,
        };
        particle.circleRange = !(0, engine_1.isObject)(angle)
            ? {
                min: minAngle,
                max: (0, engine_1.degToRad)(angle),
            }
            : { min: (0, engine_1.degToRad)(angle.min), max: (0, engine_1.degToRad)(angle.max) };
    }
}
exports.CircleDrawer = CircleDrawer;
